plugins/destination/postgresql/client/migrate.go

```diff
@@ -119,15 +119,43 @@ func (c *Client) autoMigrateTable(ctx context.Context, table *schema.Table) erro
 	}
 	if reCreatePrimaryKeys {
 		c.logger.Info().Str("table", table.Name).Msg("Recreating primary keys")
+		if err := c.setNullOnPks(ctx, table); err != nil {
+			return fmt.Errorf("failed to enforce not null on primary keys: %w", err)
+		}
+
+		tx, err := c.conn.Begin(ctx)
+		if err != nil {
+			return fmt.Errorf("failed to begin transaction to recreate primary keys: %w", err)
+		}
 		constraintName := pgx.Identifier{table.Name + "_cqpk"}.Sanitize()
 		sql := "alter table " + tableName + " drop constraint if exists " + constraintName
-		if _, err := c.conn.Exec(ctx, sql); err != nil {
+		if _, err := tx.Exec(ctx, sql); err != nil {
+			if err := tx.Rollback(ctx); err != nil {
+				c.logger.Error().Err(err).Msg("failed to rollback transaction")
+			}
 			return fmt.Errorf("failed to drop primary key constraint on table %s: %w", table.Name, err)
 		}
+		
 		sql = "alter table " + tableName + " add constraint " + constraintName + " primary key (" + strings.Join(table.PrimaryKeys(), ",") + ")"
-		if _, err := c.conn.Exec(ctx, sql); err != nil {
+		if _, err := tx.Exec(ctx, sql); err != nil {
+			if err := tx.Rollback(ctx); err != nil {
+				c.logger.Error().Err(err).Msg("failed to rollback transaction")
+			}
 			return fmt.Errorf("failed to add primary key constraint on table %s: %w", table.Name, err)
 		}
+		if err := tx.Commit(ctx); err != nil {
+			return fmt.Errorf("failed to commit transaction to recreate primary keys: %w", err)
+		}
+	}
+	return nil
+}
+
+func (c *Client) setNullOnPks(ctx context.Context, table *schema.Table) error {
```

Comment:
Should this be called `setNotNullOnPks`?

Also let's say someone changes an existing column to be a primary key and it has a null value. What will happen? Will this fail and if so, how do we handle this failure?

Reply:
Yes this will fail. we don't have (and will never have) a way to deal with that. The user will have to either drop the table or migrate columns to contain non-null.

auto-migrations can fail in data integrations platform so it's best effort and it is acceptable/industry standard as maintaining migrations for all databases*all_apis is not really feasible.